### PR TITLE
executor: fix wrong result for index merge read generate column

### LINKED ISSUE #15994: USE_INDEX_MERGE results in an incorrect result for a generated column
Consider the following statements:

```sql
CREATE TABLE t0(c0 INT AS (1), c1 INT PRIMARY KEY);
INSERT INTO t0(c1) VALUES (0);
CREATE INDEX i0 ON t0(c0);
SELECT /*+ USE_INDEX_MERGE(t0, i0, PRIMARY)*/ t0.c0 FROM t0 WHERE t0.c1 OR t0.c0;
SELECT t0.c0 FROM t0 WHERE t0.c1 OR t0.c0; -- expected: {1}, actual: {NULL}
```
Unexpectedly, an incorrect value for `c0` is returned. When removing the `USE_INDEX_MERGE` hint, the correct result is computed:
```sql
CREATE TABLE t0(c0 INT AS (1), c1 INT PRIMARY KEY);
INSERT INTO t0(c1) VALUES (0);
CREATE INDEX i0 ON t0(c0);
SELECT t0.c0 FROM t0 WHERE t0.c1 OR t0.c0; -- {1}
```
Environment:
```
mysql> select tidb_version();
+---------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------+
| tidb_version()                                                                                                                                                                                                                                                                                          |
+---------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------+
| Release Version: v4.0.0-beta.2-160-g82529c886
Git Commit Hash: 82529c886cb1cfff7810476a136acf81284bac6e
Git Branch: master
UTC Build Time: 2020-04-01 11:44:27
GoVersion: go1.13.4
Race Enabled: false
TiKV Min Version: v3.0.0-60965b006877ca7234adaced7890d7b029ed1306
Check Table Before Drop: false |
+---------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------+
1 row in set (0.00 sec)
```

### PR DESCRIPTION
**Pull request by lzmhhh123**
### What problem does this PR solve?

Issue Number: close #15994 

Problem Summary: When an index merge reader reads the generated column, it gets the wrong result. Because the generated column is ignored in the final table reader stage.

### What is changed and how it works?

What's Changed: deal the generated column for final table reader.

### Related changes

- Need to cherry-pick to the release branch

### Check List 

Tests 

- Unit test
- Integration test

Side effects

- Performance regression
    - Consumes more CPU

### CODE REVIEW

File: executor/index_merge_reader.go [hunk 1]
```diff
@@ -189,6 +189,7 @@ func (e *IndexMergeReaderExecutor) startPartialIndexWorker(ctx context.Context,
 		SetKeepOrder(false).
 		SetStreaming(e.partialStreamings[workID]).
 		SetFromSessionVars(e.ctx.GetSessionVars()).
+		SetMemTracker(e.memTracker).
```
Comment (zz-jason) [inline executor/index_merge_reader.go:192]: is this a bugfix?

Comment (lzmhhh123) [reply executor/index_merge_reader.go:192]: Probably yes. It infects the memory track for index merge reader in the index side.

File: executor/index_merge_reader_test.go [hunk 2]
```diff
@@ -50,3 +50,14 @@ func (s *testSuite1) TestJoin(c *C) {
 	tk.MustQuery("select /*+ use_index_merge(t1, t1a, t1b) */ sum(t1.a) from t1 join t2 on t1.id = t2.id where t1.a < 2 or t1.b > 4").Check(testkit.Rows("6"))
 	tk.MustQuery("select /*+ use_index_merge(t1, t1a, t1b) */ sum(t1.a) from t1 join t2 on t1.id = t2.id where t1.a < 2 or t1.b > 5").Check(testkit.Rows("1"))
 }
+
+func (s *testSuite1) TestIndexMergeReaderAndGeneratedColumn(c *C) {
+	tk := testkit.NewTestKitWithInit(c, s.store)
```
Comment (XuHuaiyu) [inline executor/index_merge_reader_test.go:55]: Do we need to `	tk.MustExec("set tidb_enable_index_merge=1")` ?

Comment (lzmhhh123) [reply executor/index_merge_reader_test.go:55]: No, the hint doesn't require `tidb_enable_index_merge`.

Comment (XuHuaiyu) [reply executor/index_merge_reader_test.go:55]: ok, but I think it's a bug.

If we do not enable_index_merge, the hint should not take effect

Comment (lzmhhh123) [reply executor/index_merge_reader_test.go:55]: Yes, I think so.

Comment (mrigger) [reply executor/index_merge_reader_test.go:55]: I don't want to interfere, but the [documentation](https://pingcap.github.io/docs/dev/reference/performance/index-merge/#enable-indexmerge) currently mentions the following:

> IndexMerge is disabled by default. Enable the IndexMerge in one of two ways:
> * Set the tidb_enable_index_merge system variable to 1;
> * Use the SQL Hint USE_INDEX_MERGE in the query.
> 
> Note: The SQL Hint has a higher priority over the system variable.

Comment (XuHuaiyu) [reply executor/index_merge_reader_test.go:55]: Alright 😢

**Pull request merged**